rs/memory_tracker/src/lib.rs

```diff
@@ -408,6 +414,58 @@ impl SigsegvMemoryTracker {
         self.accessed_bitmap.borrow().page_range()
     }
 
+    /// Converts a range of page indices into a corresponding range of memory addresses.
+    fn addr_range_from(&self, range: &Range<PageIndex>) -> Range<usize> {
+        let start_addr = self.page_start_addr_from(range.start) as usize;
+        let end_addr = self.page_start_addr_from(range.end) as usize;
+        start_addr..end_addr
+    }
+
+    /// Attempts to shrink the given page index range to be aligned with
+    /// huge page boundaries.
+    ///
+    /// If aligning either the start or end of the range would exclude
+    /// the `faulting_page`, that side of the range will remain unaligned
+    /// to ensure the faulting page stays within the adjusted range.
+    fn try_align_to_hugepage(
+        &self,
+        faulting_page: PageIndex,
+        range: Range<PageIndex>,
+    ) -> Range<PageIndex> {
+        debug_assert!(
+            range.contains(&faulting_page),
+            "Error checking page:{faulting_page} ∈ range:{range:?}"
+        );
+        let hugepage_mask = HUGE_PAGE_SIZE - 1;
+        let faulting_addr = self.page_start_addr_from(faulting_page) as usize;
+        let Range {
+            start: start_addr,
+            end: end_addr,
+        } = self.addr_range_from(&range);
+        // Try to align the range start address up to the nearest Huge Page.
+        let aligned_start_addr = (start_addr + hugepage_mask) & !hugepage_mask;
+        let start_idx = if faulting_addr >= aligned_start_addr && aligned_start_addr != start_addr {
```

Comment:
Why do we need `&& aligned_start_addr != start_addr`? Just because it'd give us `range.start` anyway and we don't want to redo the `page_index` computation in that case?

Reply:
Yes, a micro-optimization. It's all O(1) anyway, so I removed it...